{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Homework 4: Practical analysis with BioPython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the homework, you are going to extend the code from the analysis of our FASTQ file in lectures 8 and 9.\n",
    "Recall that the FASTQ file contains reads from a real sequencing run of influenza virus HA and NA genes.\n",
    "\n",
    "The _actual sequences_ are as follows:\n",
    "\n",
    "    5'-[end of HA]-AGGCGGCCGC-[16 X N barcode]-3'\n",
    "    \n",
    "or \n",
    "\n",
    "    5'-[end of NA]-AGGCGGCCGC-[16 X N barcode]-3'\n",
    "    \n",
    "The end of NA is:\n",
    "\n",
    "    ...CACGATAGATAAATAATAGTGCACCAT\n",
    "    \n",
    "The end of HA is:\n",
    "\n",
    "    ...CCGGATTTGCATATAATGATGCACCAT\n",
    "    \n",
    "The _sequencing reads_ (located in the FASTQ file) are from the reverse end of these actual sequences, so the first thing in the sequencing reads is the reverse complement of the barcode followed by the reverse complement of the constant sequence, etc.\n",
    "The reads can originate from **either** HA or NA, and that will be distinguished by the most 3' end of the read.\n",
    "But in our example exercise in class, we did not distinguish among reads matching to HA and NA, as we didn't even look far enough into the read to tell the identity.\n",
    "\n",
    "For the homework, your goal is to write code that extends the material from lectures 8 and 9 to also distinguish between HA and NA.\n",
    "This homework can be completed almost entirely by re-using code from lecture 9. You will need to set up your analysis to do the following:\n",
    " 1. Get the reverse complement of each read.\n",
    " 2. Determine if it matches the expected pattern for HA and NA, and if so which one.\n",
    " 3. If it matches, extract the barcode and add it to a dictionary to keep track of counts.\n",
    " 4. Determine the number and distribution of barcodes for HA and NA separately.\n",
    "\n",
    "Please include code to address each of the following questions. Please include code comments to explain what your code is attempting to accomplish. Don't forget to include references to the sources you used to obtain your answer, including your classmates (if you are working in groups).  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. How many reads map to HA, and how many reads map to NA?"
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There were 10000 total sequence reads parsed.\n",
      "138 reads map to NA and 281 reads map to HA.\n",
      "The total number of barcode counts that map to NA is 3910\n",
      "The total number of barcode counts that map to HA is 5246\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'AACCGTGACCAGGAAG': 70,\n",
       " 'TTATCGTCTCCCATAT': 77,\n",
       " 'CATACCAGTCATCCCT': 28,\n",
       " 'ACTTACGTATAAGTCA': 53,\n",
       " 'GCTACTACTATACCAT': 119,\n",
       " 'GTTACCCACAGTCCGC': 62,\n",
       " 'CACCACACAAGGATGT': 41,\n",
       " 'TCATACATCACACTTA': 47,\n",
       " 'GTACCCTCCGTGAATC': 99,\n",
       " 'ACTCCACGCTACCACG': 31,\n",
       " 'GCACTCCTCAACCCTT': 48,\n",
       " 'CCGCTCCCTGCTGTCC': 43,\n",
       " 'AAACGTAGCGATAACT': 61,\n",
       " 'TCACGTCCCATATTAC': 9,\n",
       " 'CCCGACCCGACATTAA': 155,\n",
       " 'ACGAGAGGTCGACTCG': 60,\n",
       " 'TTCGACTTCCTAGTAC': 86,\n",
       " 'ACCCAGTCTAGCTAAC': 70,\n",
       " 'GGTCATACGCCTTCGC': 16,\n",
       " 'CTTAACCTTCCGACAA': 55,\n",
       " 'TGGGCAATAAATGTAG': 57,\n",
       " 'CCCTCATCCTGTGTCA': 54,\n",
       " 'AATTCCATTCAGGCTG': 35,\n",
       " 'TTCTAGCCTTATCTCC': 81,\n",
       " 'AGAATAATCTCAAACT': 62,\n",
       " 'AAACAAACAAGTCTGT': 50,\n",
       " 'AATACGAACATCTCGG': 49,\n",
       " 'CGAATCTGCGCAATCT': 60,\n",
       " 'GATTTCCGATCAGTCT': 124,\n",
       " 'AGAACGTCTTGATAGC': 30,\n",
       " 'GACCGGTGCTTCAACA': 76,\n",
       " 'TGCTGACAACACGTAA': 47,\n",
       " 'ATAGTCGGGCCTACCG': 21,\n",
       " 'TTCCCAATGGTACTCG': 28,\n",
       " 'CTGGCAAGAAAATTCC': 43,\n",
       " 'TTCGGTGAACGTACAC': 76,\n",
       " 'CTCCATCACTCGCCAA': 115,\n",
       " 'TCAACCCCACGCCGCT': 24,\n",
       " 'TAACTAGCGTTTTCAA': 1,\n",
       " 'AGTACTCATCGAGCAA': 31,\n",
       " 'AACCGTCACCTCAACC': 104,\n",
       " 'GAACGACCCTTAACTA': 18,\n",
       " 'TCCAGAAACTTCAATG': 71,\n",
       " 'ACACGATACCCTTGAG': 26,\n",
       " 'CCAGCCATCCCCTCTT': 76,\n",
       " 'ATTCACACTCCTGCCA': 22,\n",
       " 'TACTTATTAGCATGTC': 46,\n",
       " 'GGAAATTATACTGCCA': 10,\n",
       " 'GCCCCACAATAACCGC': 38,\n",
       " 'GCACGCGATCTGCGAC': 32,\n",
       " 'AGGCACTAACCCAGCA': 37,\n",
       " 'ACTTCATACCATTACA': 32,\n",
       " 'CCCACCTTCACACCTA': 62,\n",
       " 'ACGAACGTCAAAAAAA': 5,\n",
       " 'GTCGCATATGAAGACG': 40,\n",
       " 'ACAAAACCCCCCGACA': 27,\n",
       " 'TCCGTATATCCAAAAG': 13,\n",
       " 'GAGCGTCCCTTCTAAA': 34,\n",
       " 'TCGCTCAGTCGTCTTT': 38,\n",
       " 'GTTACCCACAGTCGGC': 1,\n",
       " 'GAAATCCCTAACGGGC': 17,\n",
       " 'ATTTACGGGCTAGTTG': 7,\n",
       " 'AGACCGTCAGGCACCA': 46,\n",
       " 'TTCCCTAGAGACTACC': 37,\n",
       " 'TGACAACCACGTCCCC': 29,\n",
       " 'ATACTCTAAAAAATGA': 56,\n",
       " 'CACAACACGTACGCAA': 39,\n",
       " 'TGCACACACACCACGT': 38,\n",
       " 'CCCAGATCCCACCCGC': 12,\n",
       " 'GAGTCACAACCCTGCT': 1,\n",
       " 'ACTGATCAAAGCGTGG': 34,\n",
       " 'ACCGACTCACAGTACT': 52,\n",
       " 'TCGGGCGACTTAAATG': 20,\n",
       " 'GGTTCCGTGTGTGGAC': 40,\n",
       " 'GATCTCCCAAGTTGCC': 28,\n",
       " 'CGAGGCGGAAGGCGTC': 18,\n",
       " 'CGGAACGCCATCACCG': 13,\n",
       " 'TGATTTACCTAACCCG': 47,\n",
       " 'CTGGGTCCCCCCCAAA': 25,\n",
       " 'AAATGAGACAGCCTAC': 45,\n",
       " 'GAGTCACAACCCGGCT': 63,\n",
       " 'TCCCCTCACGGATTCA': 43,\n",
       " 'GTCAAGCGTCTAACCA': 23,\n",
       " 'CGCGTTAAGATCACCA': 20,\n",
       " 'CCCAGACCGTGAGCCA': 25,\n",
       " 'CAGTGATTTATTTGCG': 30,\n",
       " 'CGTCATAGAGACGGTT': 68,\n",
       " 'TGCAATACCAATAACT': 32,\n",
       " 'CTTAGTATTTCAGTAT': 81,\n",
       " 'GCTGCACCAGCTCCTC': 14,\n",
       " 'GTCGCATATGAAGACT': 4,\n",
       " 'ACACCGGTTCCTCCCC': 19,\n",
       " 'CGATCCCGTCCAGACA': 47,\n",
       " 'ACGGATAATTACCGCC': 31,\n",
       " 'CTTAACCGTCCGACAC': 1,\n",
       " 'GGAACACTGTGAAGCA': 40,\n",
       " 'CGAATCTTCGCAATCT': 1,\n",
       " 'ATTACCCCCATTTAAG': 11,\n",
       " 'GACCTCTCAGATTAGA': 42,\n",
       " 'CCAATCCCCTCGAGTG': 24,\n",
       " 'GGTTTTCTAACACCGC': 8,\n",
       " 'GTGAACTTACAGCTCC': 28,\n",
       " 'CCCCTATCATACTCTC': 5,\n",
       " 'AGGGTAGTCGACTACT': 15,\n",
       " 'GCGAAATGCAATCCTC': 26,\n",
       " 'ATAGTACCGCTCGCAA': 1,\n",
       " 'TCTTAACAGCGTCAGC': 17,\n",
       " 'CCGTACCCAAGACCAT': 12,\n",
       " 'GCGTATAAGTACGCCG': 50,\n",
       " 'GGGTCCACGTCCCAAG': 6,\n",
       " 'CGGCCCCAACGCAATA': 23,\n",
       " 'TTATACCTTAGTCAGT': 40,\n",
       " 'GGGCGTCCCTTCTAAA': 1,\n",
       " 'CCCCCCACTACAATTA': 38,\n",
       " 'GAGCACATTTATTCGT': 1,\n",
       " 'ATTGCCCGGTGGCCCT': 39,\n",
       " 'ATTGCATAACTACAAC': 73,\n",
       " 'AATTCCATTCGGGCTG': 1,\n",
       " 'GACCCCTACTTAGTGG': 23,\n",
       " 'TGCCCTGAGCCTGCAC': 8,\n",
       " 'ACGGAAACCCAGGCCA': 36,\n",
       " 'TCTGACTAGCACATCG': 18,\n",
       " 'CATCATAGAGACGGTT': 1,\n",
       " 'ACTTCATACCTTTACA': 1,\n",
       " 'GCTGCACCGGCTCCTC': 1,\n",
       " 'AATAAGAACATCTCGG': 1,\n",
       " 'CCTTTGCTTCGTAATA': 15,\n",
       " 'AAGCTTTCGTCGCGCA': 4,\n",
       " 'TGTCTAAATGTCCACG': 2,\n",
       " 'AGAGTAGTCGACTACT': 1,\n",
       " 'CCTGCATTCTGGATGC': 3,\n",
       " 'CGACGCAGCACCCCAT': 15,\n",
       " 'GACAGAGCTAGCACAG': 18,\n",
       " 'CTTAACCGTCCGACAA': 1,\n",
       " 'CGGTGCTACCTTCTAT': 1,\n",
       " 'GACCGGGGCTTCAACA': 1,\n",
       " 'GTTTCATCCCCCTATC': 18,\n",
       " 'CGGAATCAACCCCCTC': 10,\n",
       " 'TTAGAAGAAATCAAAT': 7,\n",
       " 'CATCATTCTTACTAAA': 32,\n",
       " 'CCCTGACCTAACCACG': 1,\n",
       " 'TCAGTATATCCAAAAG': 2,\n",
       " 'AGTAAATTCCCGGACG': 5,\n",
       " 'CGCACAATACTTCGAC': 35,\n",
       " 'ACAAAACCCCCCGATA': 1,\n",
       " 'TGAGTTTGCAACGAAC': 6,\n",
       " 'TTCGGTGAACGAACAC': 1,\n",
       " 'TCAACCCCACGCCTCT': 1,\n",
       " 'GACCACATTTATTCGT': 45,\n",
       " 'ACATAAATTAAAGGAA': 7,\n",
       " 'AATACGAACATCTAGG': 2,\n",
       " 'TCCACCCACTCGCGAT': 3,\n",
       " 'GCCGCAAGGAACGCTT': 14,\n",
       " 'TCCTCTAGCATCCCCA': 15,\n",
       " 'TCCAGAAACTTGAATG': 1,\n",
       " 'TCGGACGGTATACTGC': 9,\n",
       " 'TAAAAACGTTATAACA': 27,\n",
       " 'CCACCGTGGCAGCCCG': 24,\n",
       " 'CACCGTAATGTCCTAC': 7,\n",
       " 'ATACTCTAAAAAATGN': 1,\n",
       " 'TCCACCCACTCGAGAT': 1,\n",
       " 'AGGCACTAACCAAGCA': 1,\n",
       " 'TCGGGGGACTTAAATG': 1,\n",
       " 'AAACTTAATAATTTCC': 1,\n",
       " 'GGGTATAAGTACGCCG': 2,\n",
       " 'ATACCCTCCGTGAATC': 1,\n",
       " 'AAAGGTAGCGATAACT': 1,\n",
       " 'TAATACATCACACTTA': 1,\n",
       " 'ACTGATCAAAGCGTGT': 1,\n",
       " 'CATCACGCCCGTTAGC': 18,\n",
       " 'ACTCGTTGTCCCATAA': 18,\n",
       " 'GGTGATACGCCTTCGC': 1,\n",
       " 'AGAATCTGCGCAATCT': 1,\n",
       " 'CCTTTGCTTCGTAACA': 1,\n",
       " 'AAACTTACTAATTTCC': 6,\n",
       " 'CGAATCTGCGCAAACT': 1,\n",
       " 'ATTTCATAACTACAAC': 2,\n",
       " 'AGAGGCGGAAGGCGTC': 1,\n",
       " 'TTGGCCTGCCCTAAAT': 2,\n",
       " 'CGGCCCCAACGCAATG': 1,\n",
       " 'CGTCGGCCTGTTCATA': 3,\n",
       " 'TGAATTACCTAACCCG': 1,\n",
       " 'CATCACGCCCCTTAGC': 1,\n",
       " 'CTCGACCCGAGGTTAA': 1,\n",
       " 'TTTGTCCCTTGAACCC': 1,\n",
       " 'ACGGATAATTACCGGC': 1,\n",
       " 'TGGAATACCAATAACT': 1,\n",
       " 'ATGGGTCCCCCCCAAA': 1,\n",
       " 'AATACGAGCATCTCGG': 1,\n",
       " 'CGTCATAGAGAAGGTT': 1,\n",
       " 'ACGAGAGGCCGACTCG': 1,\n",
       " 'TTACTCTAAAAAATGA': 1,\n",
       " 'CGAATCTGCGCAATCC': 1,\n",
       " 'TGGGCAATAAACGTAG': 1,\n",
       " 'TTATACCTTAGTCTGT': 1,\n",
       " 'TCTTAACAGCATCAGC': 1,\n",
       " 'TTCGACTTCCTAATAC': 1,\n",
       " 'TAATATATGGATATTC': 13,\n",
       " 'CACACCAGTCATCCCT': 1,\n",
       " 'GTTAGTATTTCAGTAT': 1,\n",
       " 'GTCTCCCATCATGTAG': 2,\n",
       " 'AACCGTCACCTCTACC': 1,\n",
       " 'ACCAAGTCTAGCTAAC': 1,\n",
       " 'AACTGTGACCAGGAAG': 1,\n",
       " 'ATAGTCGGGACTACCG': 1,\n",
       " 'GCGGCAGAGCTGCTTT': 6,\n",
       " 'AAATGAGACAGACTAC': 1,\n",
       " 'GACTTACGGCGTCATG': 2,\n",
       " 'GTACACTCCGTGAATC': 1,\n",
       " 'ACACTCCTCAACCCTT': 1,\n",
       " 'TGGGAAATAAATGTAG': 1,\n",
       " 'AAAAAACCCCCCGACA': 1,\n",
       " 'CATACCTTTCTGACTT': 3,\n",
       " 'CCACCGTGTCAGCCCG': 1,\n",
       " 'TTCGAATTCCTAGTAC': 1,\n",
       " 'GATCTCTCAGATTAGA': 1,\n",
       " 'GCCCCACAATAACCAC': 1,\n",
       " 'TGGGCAAGAAATGTAG': 1,\n",
       " 'CACAACACGTACGTAA': 1,\n",
       " 'GCACGCGATCTCCGAC': 1,\n",
       " 'AACCGTCACCTCAAAC': 1,\n",
       " 'AAACAAACAAGTGTGT': 1,\n",
       " 'TATCATTCTTACTAAA': 1,\n",
       " 'TGACAACCACATCCCC': 1,\n",
       " 'ATGGAAACCCAGGCCA': 1,\n",
       " 'ATTGAATAACTACAAC': 1,\n",
       " 'AGGGGCACGAGTCGCC': 2,\n",
       " 'TCAAGAAACTTCAATG': 1,\n",
       " 'CTGGGTCGCCCCCAAA': 1,\n",
       " 'ATCGACTCACAGTACT': 1,\n",
       " 'TCCGTATAACCAAAAG': 1,\n",
       " 'GGTCATACGCCTCCGC': 1,\n",
       " 'ACAGGCAGCTACGGGA': 3,\n",
       " 'AGAATAATCTCAGACT': 1,\n",
       " 'TTATCGTCTCCCATAC': 1,\n",
       " 'CTTAGTAGTTCAGTAT': 1,\n",
       " 'TTCGGTGAAAGTACAC': 1,\n",
       " 'GACCACATTTATCCGT': 1,\n",
       " 'CCGCTCCCTCCTGTCC': 1,\n",
       " 'AAACGTGACCAGGAAG': 1,\n",
       " 'GACACTCCAGTCGACA': 1,\n",
       " 'CCACCGTGGGAGCCCG': 1,\n",
       " 'ACCGACTGACAGTACT': 1,\n",
       " 'TGCTGACAACGCGTAA': 1,\n",
       " 'GCTGCACCATCTCCTC': 1,\n",
       " 'ACCAGACCGTGAGCCA': 1,\n",
       " 'GGAAATTATACTCCCA': 1,\n",
       " 'TGCAATACCAATAACG': 1,\n",
       " 'TATTTCCGATCAGTCT': 1,\n",
       " 'GATATAGCTAGCACAG': 1,\n",
       " 'GTTTCATACCCCTATC': 1,\n",
       " 'GTACCCTACGTGAATC': 1,\n",
       " 'GTGAACTTAAAGCTCC': 1,\n",
       " 'GTCCCATATGAAGACG': 1,\n",
       " 'AGTACTCATCGTGCAA': 1,\n",
       " 'CCCAGACCGTGAGACA': 1,\n",
       " 'CTCCTTCACTCGCCAA': 1,\n",
       " 'CGAATCTGCGCAATAT': 1,\n",
       " 'GCGAAATGCAATCCCC': 1,\n",
       " 'CCCACCTTCACAACTA': 1,\n",
       " 'AGCACACCCAGCTCGA': 1,\n",
       " 'AGACCGTCAGACACCA': 1,\n",
       " 'CGAGGCGGAAGGGGTC': 1,\n",
       " 'GTACCATCCGTGAATC': 1,\n",
       " 'CAAGCCATCCCCTCTT': 1,\n",
       " 'TTCCCTAGAGACTAAC': 1,\n",
       " 'TCTTAACATCGTCAGC': 1,\n",
       " 'CGAATCTGAGCAATCT': 1,\n",
       " 'ACCGACCAGACATTAA': 1,\n",
       " 'ATTCACACTACTGCCA': 1,\n",
       " 'CCGCTCCCTGCTGTAC': 1,\n",
       " 'TACGGTGAACGTACAC': 1,\n",
       " 'CCCTACCCGACATTAA': 1,\n",
       " 'TGACAAACACGTCCCC': 1,\n",
       " 'AGGCCCCAACGCAATA': 1,\n",
       " 'ATAGTCGAGCCTACCG': 1,\n",
       " 'ACCGACGCACAGTACT': 1,\n",
       " 'GCACGCGATTTGCGAC': 1,\n",
       " 'ATTGCATGACTACAAC': 1,\n",
       " 'AATACGAACATATCGG': 1,\n",
       " 'GCTACTACTATACCTT': 1}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load necessary packages\n",
    "import re\n",
    "import Bio.SeqIO\n",
    "\n",
    "#Read sequences from fastq file\n",
    "seqreads = list(Bio.SeqIO.parse('barcodes_R1.fastq', 'fastq')) \n",
    "seqreads_str = []\n",
    "for seqrecord in seqreads:\n",
    "    seqreads_str.append(str(seqrecord.seq))\n",
    "\n",
    "#reverse complement of sequence reads\n",
    "def reverse_complement(seq, unk_partner = 'N'):\n",
    "    base_partner = {'A':'T', 'T':'A', 'C':'G', 'G':'C'}\n",
    "    rseq = ''\n",
    "    for a in seq:\n",
    "        if a in base_partner: # look up the complementary base in the dictionary\n",
    "            pair = base_partner[a]\n",
    "            rseq = pair + rseq\n",
    "        else:\n",
    "            rseq = unk_partner + rseq\n",
    "    return rseq\n",
    "\n",
    "#create 2 functions -- one that will be for NA reads with NA end and another that will read HA reads with HA end\n",
    "#function for NA reads\n",
    "def NA_read_barcode(seqread, bclen, upstream='AGGCGGCCGC', NAend='CACGATAGATAAATAATAGTGCACCAT'):\n",
    "    reverse = reverse_complement(seqread) # get the reverse complement of the read\n",
    "    \n",
    "    NAend_pattern = re.compile(NAend + upstream + f\"(?P<NAbarcode>[ATCGN]{{{bclen}}})$\") # compile the barcode search pattern that has end of NA sequence + upstream seq + 16xN barcode\n",
    "    NA_match = NAend_pattern.search(reverse) # search for the barcode pattern that has end of HA sequence\n",
    "\n",
    "    if NA_match:\n",
    "        NAbarcode = NA_match.group('NAbarcode')\n",
    "        return NAbarcode\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "#for HA reads\n",
    "def HA_read_barcode(seqread, bclen, upstream='AGGCGGCCGC', HAend='CCGGATTTGCATATAATGATGCACCAT'):\n",
    "    reverse = reverse_complement(seqread) # get the reverse complement of the read\n",
    "\n",
    "    HAend_pattern = re.compile(HAend + upstream + f\"(?P<HAbarcode>[ATCGN]{{{bclen}}})$\") # compile the barcode search pattern that has end of HA sequence + upstream seq + 16xN barcode\n",
    "    HA_match = HAend_pattern.search(reverse) # search for the barcode pattern that has end of HA sequence\n",
    "\n",
    "    if HA_match:\n",
    "        HAbarcode = HA_match.group('HAbarcode')\n",
    "        return HAbarcode\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "#get counts of HA and NA but first make empty dictionaries for each \n",
    "NA_barcode_counts = {}\n",
    "HA_barcode_counts = {}\n",
    "\n",
    "for seq in seqreads_str: # iterate through all reads\n",
    "    NAbarcode = NA_read_barcode(seq, bclen=16, upstream='AGGCGGCCGC', NAend='CACGATAGATAAATAATAGTGCACCAT') # NAbarcode variable stores NA_read_barcode functioning and set bclen length=16\n",
    "    if NAbarcode: # if there is a valid barcode for NA, add it to the empty dictionary called NA_barcode_counts\n",
    "        if NAbarcode in NA_barcode_counts:\n",
    "            NA_barcode_counts[NAbarcode] += 1\n",
    "        else:\n",
    "            NA_barcode_counts[NAbarcode] = 1\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "for seq in seqreads_str: # iterate through all reads\n",
    "    HAbarcode = HA_read_barcode(seq, bclen=16, upstream='AGGCGGCCGC', HAend='CCGGATTTGCATATAATGATGCACCAT') # NAbarcode variable stores HA_read_barcode functioning and set bclen length=16\n",
    "    if HAbarcode: # if there is a valid barcode for NA, add it to the empty dictionary called HA_barcode_counts\n",
    "        if HAbarcode in HA_barcode_counts:\n",
    "            HA_barcode_counts[HAbarcode] += 1 \n",
    "        else:\n",
    "            HA_barcode_counts[HAbarcode] = 1\n",
    "\n",
    "\n",
    "\n",
    "print(f\"There were {len(seqreads_str)} total sequence reads parsed.\") #total sequence reads\n",
    "print(f\"{len(NA_barcode_counts)} reads map to NA and {len(HA_barcode_counts)} reads map to HA.\") #mapped reads to HA and NA\n",
    "\n",
    "print(\"The total number of barcode counts that map to NA is \" + str(sum(NA_barcode_counts.values()))) #total barcode counts for NA\n",
    "print(\"The total number of barcode counts that map to HA is \" + str(sum(HA_barcode_counts.values()))) #total barcode counts for HA\n",
    "\n",
    "NA_barcode_counts #defined function that gives dictionary key and values for NA_barcodes and their counts\n",
    "HA_barcode_counts #defined function that gives dictionary key and values for HA_barcodes and their counts\n"
   ]
=======
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
>>>>>>> 609ed814557ab89799e45a04e39e70609cb8cc7a
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. What is the HA barcode with the most counts (and how many counts)? Also answer the same question for NA.\n",
    "\n",
    "    _Hint: you will need to find the key associated with the maximum value in your dictionary. There are many ways to do this._"
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The barcode TTTCTCCCGCATCTCG has the most NA counts with 152\n",
      "The barcode TTTGTCCCTTGAACCC has the most HA counts with 155\n"
     ]
    }
   ],
   "source": [
    "NAmax_value = max(NA_barcode_counts) # NA barcode with the most counts\n",
    "NAmax_value_counts = max(NA_barcode_counts.values()) # counts for the NA barcode with the most counts\n",
    "print(f\"The barcode {NAmax_value} has the most NA counts with {NAmax_value_counts}\")\n",
    "\n",
    "HAmax_value = max(HA_barcode_counts) # NA barcode with the most counts\n",
    "HAmax_value_counts = max(HA_barcode_counts.values()) # counts for the NA barcode with the most counts\n",
    "print(f\"The barcode {HAmax_value} has the most HA counts with {HAmax_value_counts}\")"
   ]
  },
  {
   "cell_type": "code",
=======
>>>>>>> 609ed814557ab89799e45a04e39e70609cb8cc7a
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {
    "height": "623.8px",
    "left": "0px",
    "right": "829.6px",
    "top": "33.2px",
    "width": "322.4px"
   },
   "toc_section_display": "block",
   "toc_window_display": true
  },
  "vscode": {
   "interpreter": {
    "hash": "d4d1e4263499bec80672ea0156c357c1ee493ec2b1c70f0acce89fc37c4a6abe"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
